src/thor/expansion_action.cc

```diff
@@ -0,0 +1,152 @@
+#include "midgard/logging.h"
+#include "midgard/polyline2.h"
+#include "midgard/util.h"
+
+using namespace rapidjson;
+using namespace valhalla::midgard;
+
+namespace {
+void FillDom(Document& dom, valhalla::Options::Action action) {
+  SetValueByPointer(dom, "/type", "FeatureCollection");
+  SetValueByPointer(dom, "/features/0/type", "Feature");
+  SetValueByPointer(dom, "/features/0/geometry/type", "MultiLineString");
+  SetValueByPointer(dom, "/features/0/geometry/coordinates", Value(kArrayType));
+  if (action == valhalla::Options::route) {
+    SetValueByPointer(dom, "/properties/algorithm", "none");
+    SetValueByPointer(dom, "/features/0/properties/edge_ids", Value(kArrayType));
+    SetValueByPointer(dom, "/features/0/properties/statuses", Value(kArrayType));
+  } else {
+    SetValueByPointer(dom, "/features/0/properties/durations", Value(kArrayType));
+    SetValueByPointer(dom, "/features/0/properties/distances", Value(kArrayType));
+    SetValueByPointer(dom, "/features/0/properties/costs", Value(kArrayType));
+  }
+}
+} // namespace
+
+namespace valhalla {
+namespace thor {
+
+std::string thor_worker_t::expansion(Api& request) {
+  // time this whole method and save that statistic
+  measure_scope_time(request, "thor_worker_t::expansion");
+
+  auto options = request.options();
+  auto exp_action = options.expansion_action();
+  // default generalization to ~ zoom level 15
+  float gen_factor = options.has_generalize() ? options.generalize() : 21.4f;
+
+  // default the expansion geojson so its easy to add to as we go
+  Document dom;
+  dom.SetObject();
+  FillDom(dom, exp_action);
+
+  std::cout << std::setw(6) << "Name" << std::setw(10) << "dir" << std::setw(8) << "dur"
+            << std::setw(8) << "dist" << std::setw(8) << "cost" << std::endl;
+
+  // a lambda that the path algorithm can call to add stuff to the dom
+  // route and isochrone produce different GeoJSON properties
+  auto track_expansion = [&dom, &exp_action,
+                          &gen_factor](baldr::GraphReader& reader, baldr::GraphId edgeid,
+                                       const char* algorithm = nullptr, const char* status = nullptr,
+                                       const float duration = 0, const float distance = 0,
+                                       const float cost = 0) {
+    auto tile = reader.GetGraphTile(edgeid);
+    if (tile == nullptr) {
+      LOG_ERROR("thor_worker_t::expansion error, tile no longer available" +
+                std::to_string(edgeid.Tile_Base()));
+      return;
+    }
+    const auto* edge = tile->directededge(edgeid);
+    auto shape = tile->edgeinfo(edge).shape();
+
+    // TODO: remove debug stuff
+    std::cout << std::setprecision(4) << std::setw(6) << tile->edgeinfo(edge).GetNames()[0]
+              << std::setw(10) << (edge->forward() ? ": forward" : ": reverse") << std::setw(8)
+              << duration << std::setw(8) << distance << std::setw(8) << cost << std::endl;
```

Comment:
will print a table like (here part of the included test output):

```
  Name       dir     dur    dist    cost
    AD : forward    0.96      20   0.816
    AB : forward    0.96      20   0.816
    AB : reverse   29.42      40   1.632  -- higher cost than next one?
    AD : reverse   29.42      40   1.632  -- higher cost than next one?
    AE : forward    1.44      30   1.224
    AC : forward    1.44      30   1.224
    ...
```

Bit surprised that smth with higher cost and u-turn (`AB`&`AD` reverse) is popped off the queue before smth with lower cost (`AE`&`AC` forward).. so sort cost is smth else than cost.cost?